akka-projection-grpc/src/main/scala/akka/projection/grpc/consumer/scaladsl/EventConsumer.scala

```diff
@@ -0,0 +1,202 @@
+          slice = eventEnvelope.slice,
+          filtered = eventEnvelope.filtered,
+          source = eventEnvelope.source,
+          tags = newTags)
+    }
+
+  }
+
+  private final case class MapPersistenceId(f: EventEnvelope[Any] => String) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newPid = f(eventEnvelope)
+      if (newPid eq eventEnvelope.persistenceId) eventEnvelope
+      else
+        new EventEnvelope[Any](
+          offset = eventEnvelope.offset,
+          persistenceId = newPid,
+          sequenceNr = eventEnvelope.sequenceNr,
+          eventOption = eventEnvelope.eventOption,
+          timestamp = eventEnvelope.timestamp,
+          eventMetadata = eventEnvelope.eventMetadata,
+          entityType = eventEnvelope.entityType,
+          slice = eventEnvelope.slice,
+          filtered = eventEnvelope.filtered,
+          source = eventEnvelope.source,
+          tags = eventEnvelope.tags)
+    }
+  }
+
+  private final case class MapPayload(f: EventEnvelope[Any] => Option[Any]) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] = {
+      val newMaybePayload = f(eventEnvelope)
+      if (newMaybePayload eq eventEnvelope.eventOption) eventEnvelope
+      else {
+        new EventEnvelope[Any](
+          offset = eventEnvelope.offset,
+          persistenceId = eventEnvelope.persistenceId,
+          sequenceNr = eventEnvelope.sequenceNr,
+          eventOption = newMaybePayload,
+          timestamp = eventEnvelope.timestamp,
+          eventMetadata = eventEnvelope.eventMetadata,
+          entityType = eventEnvelope.entityType,
+          slice = eventEnvelope.slice,
+          filtered = newMaybePayload.isEmpty,
+          source = eventEnvelope.source,
+          tags = eventEnvelope.tags)
+      }
+
+    }
+  }
+
+  private final case class AndThen(first: Transformation, next: Transformation) extends Transformation {
+    override private[akka] def apply(eventEnvelope: EventEnvelope[Any]): EventEnvelope[Any] =
+      next.apply(first.apply(eventEnvelope))
+  }
+
+  /**
+   * Transformation of events from the producer type to the internal representation stored in the journal
+   * and seen by local projections.
+   *
+   * Events can be excluded by mapping them to `None`.
+   */
+  sealed trait Transformation {
+    def mapTags[Event](f: EventEnvelope[Event] => Set[String]): Transformation =
+      MapTags(f.asInstanceOf[EventEnvelope[Any] => Set[String]])
+    def mapPersistenceId[Event](f: EventEnvelope[Event] => String): Transformation =
+      MapPersistenceId(f.asInstanceOf[EventEnvelope[Any] => String])
+    def mapPayload[Event, B <: Event](f: EventEnvelope[Event] => Option[B]): Transformation =
+      MapPayload(f.asInstanceOf[EventEnvelope[Any] => Option[B]])
+
+    def andThen(transformation: Transformation): Transformation =
```

Comment:
I think it's unfortunate if we have a completely different style of API for this consumer.Transformation compared to the producer.Transformation. Can't we make it look more similar to the existing, with such `registerX` methods?

Reply:
I agree, I'll align.

Reply:
Aligned names, starting point `empty` and composing in [658c1de](https://github.com/akka/akka-projection/pull/932/commits/658c1de80d8de732384a51bfb09676e8a0738b0c)

I didn't go the whole way and make it match event type the same way, with a class tag everywhere, do you think that should be the same as well?